classes/class-form-generator.php

```diff
@@ -0,0 +1,200 @@
+					'name'        => '',
+					'value'       => '',
+					'options'     => array(),
+					'classes'     => '',
+					'data'        => array(),
+				) );
+
+				$current_value = $args['value'];
+
+				$output  = sprintf(
+					'<select name="%1$s" class="%2$s" id="%1$s">',
+					esc_attr( $args['name'] ),
+					esc_attr( $args['classes'] )
+				);
+
+				foreach ( $args['options'] as $value => $label ) {
+					$output .= sprintf(
+						'<option value="%1$s" %2$s>%3$s</option>',
+						esc_attr( $value ),
+						selected( $value === $current_value, true, false ),
+						esc_html( $label )
+					);
+				}
+				$output .= '</select>';
+				break;
+			case 'select2':
+				$args = wp_parse_args( $original_args, array(
+					'name'        => '',
+					'value'       => '',
+					'options'     => array(),
+					'classes'     => '',
+					'data'        => array(),
+				) );
+
+				$values = array();
+
+				$output = sprintf(
+					'<select name="%1$s" id="%1$s" class="select2-select %2$s" %3$s/>',
+					esc_attr( $args['name'] ),
+					esc_attr( $args['classes'] ),
+					$this->prepare_data_string( $args['data'] )
+				);
+
+				foreach ( $args['options'] as $parent ) {
+					$parent = wp_parse_args( $parent, array(
+						'value'    => '',
+						'text'     => '',
+						'children' => array(),
+					) );
+					if ( ! empty( $parent['children'] ) ) {
+						$output .= sprintf(
+							'<optgroup label="%1$s" value="%2$s">',
+							$parent['text'],
+							$parent['value']
+						);
+						foreach ( $parent['children'] as $child ) {
+							$output .= sprintf(
+								'<option value="%1$s" %3$s>%2$s</option>',
+								$child['value'],
+								$child['text'],
+								selected( $args['value'], $child['value'], false )
+							);
+							$values[] = $child['value'];
+						}
+						$output .= '</optgroup>';
+					} else {
+						$output .= sprintf(
+							'<option value="%1$s" %3$s>%2$s</option>',
+							$parent['value'],
+							$parent['text'],
+							selected( $args['value'], $parent['value'], false )
+						);
+						$values[] = $parent['value'];
+					}
+				}
+
+				if ( ! empty( $args['value'] ) && ! in_array( $args['value'], $values, true ) ) {
+					$output .= sprintf(
+						'<option value="%1$s" %2$s>%1$s</option>',
+						$args['value'],
+						selected( true, true, false )
+					);
+				}
+
+				$output .= '</select>';
+				break;
+			case 'checkbox':
+				$output = sprintf(
+					'<input type="checkbox" name="%1$s" id="%1$s" value="1" %3$s>%2$s',
+					$args['name'],
+					$args['text'],
+					checked( $args['value'], true, false )
+				);
+				break;
+			default:
+				$output = apply_filters( 'wp_stream_form_render_field', $output, $field_type, $original_args );
+				break;
+		}
+
+		$output .= ! empty( $args['description'] ) ? wp_kses_post( sprintf( '<p class="description">%s</p>', $args['description'] ) ) : null;
+
+		return $output;
+	}
+
+	public function prepare_data_string( $data ) {
```

Comment:
This method prepares the HTML data attributes key/value pairs. For instance, if you have an array with `array( 'my-key' => 'my-value' )` it will turn it into an HTML data attribute: `data-my-key="my-value"`. 

This is used to output all of the data attributes on the form elements.